### ISSUE TITLE
zip -e bug

### ISSUE THREAD

**Issue by pmqs** [pmqs/z1#285]
### Posted by wiz » Sun May 06, 2012 6:25 am
`"zip -e foo bar` adds a (short) encrypted file to a zip archive
correctly. However, it sets bit 3 in the general purpose bit flags
(meaning there should be a data descriptor after the compressed data)
but still sets the crc/size/compressed size fields BEFORE the
compressed data correctly.

The PKZIP appnote says:

> Bit 3: If this bit is set, the fields crc-32, compressed
size and uncompressed size are set to zero in the
local header. The correct values are put in the
data descriptor immediately following the compressed
data. (Note: PKZIP version 2.04g for DOS only
recognizes this bit for method 8 compression, newer
versions of PKZIP recognize this bit for any
compression method.)

In my reading of the text, it should set size/compressed size/crc
before the compressed data to zero; or just not add a data descriptor.
I'm not sure why it does add it, because both input and output files are seekable,
so it can just write the values in the standard fields instead.

If there is a reason to write the data descriptor, then I think that the values before the data should be set to zero like the appnote.txt says.

Here's my `zip -v` output, in case it matters:

```
Copyright (c) 1990-2008 Info-ZIP - Type 'zip "-L"' for software license.
This is Zip 3.0 (July 5th 2008), by Info-ZIP.
Currently maintained by E. Gordon. Please send bug reports to
the authors using the web page at http://www.info-zip.org; see README for details.

Latest sources and executables are at ftp://ftp.info-zip.org/pub/infozip,
as of above date; see http://www.info-zip.org/ for other sites.

Compiled with gcc 4.5.3 for Unix (NetBSD 6.99.4) on Mar 22 2012.

Zip special compilation options:
USE_EF_UT_TIME (store Universal Time)
BZIP2_SUPPORT (bzip2 library version 1.0.5, 10-Dec-2007)
bzip2 code and library copyright (c) Julian R Seward
(See the bzip2 license for terms of use)
SYMLINK_SUPPORT (symbolic links supported)
LARGE_FILE_SUPPORT (can read and write large files on file system)
ZIP64_SUPPORT (use Zip64 to store large files in archives)
UNICODE_SUPPORT (store and read UTF-8 Unicode paths)
STORE_UNIX_UIDs_GIDs (store UID/GID sizes/val<FC>s using new extra field)
UIDGID_NOT_16BIT (old Unix 16-bit UID/GID extra field not used)
[encryption, version 2.91 of 05 Jan 2007] (modified for Zip 3)

Encryption notice:
The encryption code of this program is not copyrighted and is
put in the public domain. It was originally written in Europe
and, to the best of our knowledge, can be freely distributed
in both source and object forms from any country, including
the USA under License Exception TSU of the U.S. Export
Administration Regulations (section 740.13(e)) of 6 June 2002.

Zip environment options:
ZIP: [none]
ZIPOPT: [none]
```

**Comment by pmqs** [comment]
### Posted by E G » Thu May 10, 2012 2:01 am
It's complicated, but a reply should be coming shortly. Stay tuned.

**Comment by pmqs** [comment]
### Posted by E G » Fri May 11, 2012 11:29 pm
Traditional encryption (the original PKZip encryption algorithm) uses the CRC of the file as part of the encryption header information and that is used to actually encrypt the file. However, Zip doesn't always know what the CRC is when starting to process the file. It does if the file is coming from another archive, for instance, but typically at the point the header is written the file has not been read yet. Usually this is just a question of efficiency (we don't want to read the file once to get the CRC and then again to process the data), but this is especially a problem when streaming, where the data can't be determined in advance. Usually Zip will calculate the CRC after the data is read and then back up and update the CRC, but when using traditional encryption, the header has already been used to actually encrypt the data and so can't be changed.

The way Info-ZIP solved this chicken and egg problem is by using the modification date/time as the CRC. Though the note in zipup.c says that this is temporary, it actually ends up being used to encrypt the file and so the entry is stuck with using it. This works fine, but then there needs to be a place to store the real CRC. This ends up being the data descriptor. So even though the size information in the data descriptor is not needed (is correct in the actual entry), the CRC is needed. Given there is a data descriptor, there is no need for storing size information in the local header, but this has traditionally been done in this case and so we continue to do it. Technically this information should be ignored if there is a data descriptor, so what's in the local header doesn't really matter. Since we have the correct numbers, we might as well use them.

Note that ZipCloak only works on existing entries in existing archives, and so knows the CRC and does not need to use data descriptors. However, to remove an entry's dependence on the data descriptor, the entry needs to be decrypted, then encrypted again. Encrypted entries are otherwise just copied as is, including the data descriptors if they exist. Zip will generally remove data descriptors when copying entries from one archive to another, but encrypted entries are the exception and must remain unchanged if they are to be decrypted later.

That was a rather good question. Let me know if my answer is lacking or you have additional inquiries.

-Ed

**Comment by pmqs** [comment]
### Posted by sms » Sat May 12, 2012 11:19 pm
That sounds plausible, but if I do `zip -e` on a small text file, I
seem to get a local file header with these data:
```
CRC: C8 D5 DD F8 CSize: 00 00 00 18 USize: 00 00 00 1C
```
followed by a data descriptor with these data:
```
CRC: C8 D5 DD F8 CSize: 00 00 00 18 USize: 00 00 00 1C
```
and a central directory header with these data:
```
CRC: C8 D5 DD F8 CSize: 00 00 00 18 USize: 00 00 00 1C
```

It's not immediately obvious what the data descriptor is adding.

http://antinode.info/ftp/info-zip/forum/encry_31.zip

(Password = "password".)

**Comment by pmqs** [comment]
### Posted by E G » Sun May 13, 2012 6:50 am
Let me trace it through again. Maybe I missed something. (Or something has changed somewhere since I looked at it, years ago.)

As an exercise, try streaming and see what you get.

**Comment by pmqs** [comment]
### Posted by sms » Wed May 16, 2012 3:34 pm
> As an exercise, try streaming and see what you get.

Fun and education?

My first (too-)simple attempt at a test of streaming failed, by not
actually streaming. Apparently, the shell (bash on HP-UX) was clever
enough to take a command like `zip - [...] > archive.zip`, and connect
`archive.zip` directly to the program, so it's still seen as seekable.
A command like `zip - [...] | cat > archive.zip` solved that problem.

So much for education. Now, on to fun.

With actual streaming, both with and without encryption (traditional
is all I tried), we seem to put out a local header with an accurate
uncompressed size (not zero), but with zero for the compressed size.
Without encryption, the local header crc is zero, but with encryption,
it's not zero (DOS mod time in the two high bytes?). Then, as expected,
the data descriptor and the central directory header have all-correct
data.

When streaming (and thus using a data descriptor), putting an
accurate uncompressed size into the local header seems to violate the
APPNOTE (as cited in the original complaint), although it's not clear
that this should cause any actual problems.

With a seekable archive, we seem to put out a local header with
all-correct data, but with (traditional) encryption, we also put out a
data descriptor, which is apparently redundant.

Just in case, I tried using a file large enough not to fit in Zip's
internal buffers, but the results were unchanged.

Another interesting topic is what happens when archiving stdin ("-").
In this case, we put out a local header with -1 for the two 32-bit
(uncompressed and compressed) sizes, and then a Zip64 extra block with
the two 64-bit sizes. But, of course, these 64-bit sizes are all
zero/useless. In the APPNOTE, the `Zip64 Extended Information Extra
Field (0x0001)` description does say, "This entry in the Local header
must include BOTH original and compressed file size fields", but it
seems as if being able to omit the two "must include" sizes would save
16 bytes with no loss of useful info. (Wouldn't an empty Zip64 extra
block still convey the 64-bitness of the sizes in the inevitable data
descriptor? What else is needed?)

**Comment by pmqs** [comment]
### Posted by sms » Wed May 16, 2012 6:50 pm
> Traditional encryption (the original PKZip encryption algorithm) uses
> the CRC of the file as part of the encryption header information and
> that is used to actually encrypt the file. [...]

I wouldn't trust me, but, ...

As I read it, `crypt.c:crypthead()` seems to be using a bunch of
(pseudo-)random data to construct a "random header" for use by the
(traditional) encryption algorithm. Most of these data are results from
`rand()`, but the last two bytes are taken from the high two bytes of the
`(int) "crc"` argument passed in by the caller. (So it makes some sense
that the fake CRC value is formed by putting the DOS mod time into its
high 16 bits. `zipup.c:zipup()` does: `z->crc = z->tim << 16;`, and then
passes `z->crc` into `crypthead()`.)

But this just uses (half of) the (fake) CRC as a convenient source of
"random" data for the encryption algorithm, not as any actual CRC which
will be compared against any other CRC at any place or time. Any other
two non-CRC random bytes would serve the same purpose. Thus, there's no
good reason for zipup.c:zipup() to save the fake CRC as if it were a
real CRC (`z->crc = z->tim << 16;`). Simply calling `crypt.c:crypthead()`
with the appropriate (fake CRC) argument should work just as well:

```diff
< crypthead(key, z->crc);
---
> crypthead(key, z->tim << 16);
```
So, I still see no reason to use a data descriptor with encryption in
a seekable archive.

**Comment by pmqs** [comment]
### Posted by sms » Wed Oct 17, 2012 11:31 pm
> So, I still see no reason to use a data descriptor with encryption in
> a seekable archive.

I recently ran an experiment using code modified not to use data
descriptors on encrypted files (with a seekable archive), and all seemed
to work fine (as expected). The code change looked like this (in the
latest development code, but probably similar in any recent kit):
Code: Select all

```diff
--- zipup.c;-2  2012-08-02 17:48:28 -0500
+++ zipup.c     2012-10-17 15:55:04 -0500
@@ -1602,6 +1602,8 @@
         z->ver = 46; break;
 #endif
       }
+/* SMSd. */
+#ifndef CRYPT_NOEXT
       /*
        * The encryption header needs the crc, but we don't have it
        * for a new file.  The file time is used instead and the encryption
@@ -1614,6 +1616,7 @@
         /* not encrypting so don't need extended local header */
         z->flg &= ~8;
       }
+#endif /* ndef CRYPT_NOEXT */

 #ifdef CRYPT_AES_WG
       if (z->encrypt_method > 1) {
@@ -1639,6 +1642,8 @@
       if (zfseeko(y, bytes_this_split, SEEK_SET))
         return ZE_READ;

+/* SMSd. */
+#ifndef CRYPT_NOEXT
       if ((z->flg & 1) != 0) {
 #ifdef CRYPT_AES_WG
         if (z->encrypt_method == 1)
@@ -1657,6 +1662,7 @@
         tempzn += 16L;
 #endif
       }
+#endif /* ndef CRYPT_NOEXT */
     }
   } /* isdir */
   /* Free the local extra field which is no longer needed */
```
And the build was done with `LOCAL_ZIP=-DCRYPT_NOEXT`` (or similar).

On an HP-UX system (where it's easier to get file byte counts than it
is on my usual VMS system), I created two (traditionally encrypted)
archives of a recent Zip development tree, one with an older Zip (which
added the data descriptors), and one with the modified Zip (which
didn't add them):
Code: Select all
```
-rw-r--r--   1 root       sys        4366184 Oct 17 06:48 zip_s9.zip
-rw-r--r--   1 root       sys        4355704 Oct 17 06:49 zip_sA.zip
```
The archive size reduction, 4366184 - 4355704 = 10480 = 16 * 655,
agrees with the actual file counts in the archives:

```
dyi# ../unzip610c08a_l_sB_gcc/unzip -Z -t zip_s9.zip
Members: 722; Bytes uncompressed: 14003117, compressed: 4204778, 70.0%
Directories: 67, Files: 655, Links: 0
```

When I piped the Zip output, making it non-seekable, the resulting
archive was equivalent to one produced by the old Zip. (It still has a
non-zero (accurate) byte count in a local header, which seems to violate
a strict interpretation of the APPNOTE, but this doesn't bother our
UnZip, and it seems likely to be harmless elsewhere, so I'm less eager
to change that behavior.)

Until someone provides a good counter-argument, I'd suggest ripping
this code out of zipup.c, and using data descriptors only where they're
really needed.

**Comment by pmqs** [comment]
### Posted by sms » Tue Oct 30, 2012 10:55 pm
> [...] The code change looked like this [...]

It's a little more complicated than that, of course. In zipup.c, the
code which sets the flag bit and the code which decides whether to use
putextended() to put out the extended header (data descriptor) use
different criteria. It's not immediately clear that putting out an
extended header after a zfseeko() failure makes sense, when the
corresponding flag bit has not been set. My latest development code is
more careful about setting the flag bit:

```C
  if (use_descriptors)
    z->flg |= 8;
```
but I haven't done anything to stop using putextended() in cases where
the flag bit has not been set.

While working on fUnZip, I got some errors when the flag bit was set,
but the extended header (data descriptor) was missing. I don't know if
anyone would care if the flag bit were clear, but the extended header
appeared, anyway. Unless someone knows that it's safe, I'd suggest
changing the logic to set the flag bit as desired, and then letting the
flag bit (only) determine whether putextended() gets called (and not
caring (explicitly) about zfseeko() failures, use_descriptors, and so
on).

**Comment by pmqs** [comment]
### Posted by sms » Thu Mar 07, 2013 4:12 pm
> So, I still see no reason to use a data descriptor with encryption in
> a seekable archive.

I do now. With a data descriptor, UnZip expects one pattern in the
encrypted data, and without a data descriptor, UnZip expects another
pattern in the encrypted data. Each method stores two bytes of
random-looking data in the encryption header data, but they don't use
the same data. Sadly, it's much easier and more efficient to make the
with-descriptor pattern.

The PKWARE APPNOTE describes the normal (no-descriptor) method, and
it includes part of the CRC-32 checksum of the file in the encryption
header data. Currently, Zip does not know that checksum until it has
processed a file, and then it's too late to change the encryption header
data. In principle, Zip could process a file twice, once to calculate
the checksum, and again to add it to the archive, but this would be
extra work. (Beside the actual file reading, Zip does the line-ending
adjustment and the EBCDIC-to-ASCII conversion before calculating the
checksum. See `zipup.c:iz_file_read()`.) For large files, the extra time
might be a significant burden. For input from a pipeline, processing
the input twice would require saving it somewhere.

The with-descriptor method includes the MS-DOS-format modification
time of a file in the encryption header data (instead of its CRC-32
checksum), and this is known before the file is processed, which allows
Zip to process a file only once.

When UnZip decrypts the encryption header data, it checks these two
bytes in the encryption header against the appropriate value, and it
uses the presence of the data descriptor to decide which value (checksum
or mod-time) is the appropriate value.

If there is some clever way to re-write the encryption header _after_
processing (and encrypting) a file (so that the checksum is known) such
that the normal encryption algorithm still works, then we don't know
what it is. Without that ability, it seems to make more sense to pay
the small price of adding the data descriptor than it does to add the
extra complication and work required to avoid adding it.

**Comment by pmqs** [comment]
### Posted by E G » Wed Jun 25, 2014 3:06 am
I see in your last post you arrived at why descriptors are used with traditional encryption -- it is to flag the encryption approach used among other things. Note that WinZip has figured this out and their code is compatible with the way Zip does it. Changing our code at this point might make our traditional encryption entries incompatible with WinZip.

It looks like Zip 3.1 will include the "ETWODD" code you developed that does a second read on a file to determine the CRC in advance and so remove the need for a data descriptor. However, other than that, I wouldn't change too much at this point, unless we somehow can create non-data descriptor entries as fast as data descriptor traditional encryption entries.

**Comment by pmqs** [comment]
### Posted by E G » Fri Nov 27, 2015 8:41 am
Public Beta Zip 3.1d has posted. See the Releases forum thread for details.
Top

**Comment by pmqs** [comment]
>[!NOTE]
The text in this issue was sourced from [here](https://web.archive.org/web/20180203061406/http://www.info-zip.org/phpBB3/viewtopic.php?f=6&t=388&sid=5dd033a7c85d778ddf7f3cfa90f7e7ff)
